src/librustc_codegen_llvm/type_of.rs

```diff
@@ -52,32 +52,36 @@ fn uncached_llvm_type<'a, 'tcx>(
     }
 
     let name = match layout.ty.kind {
-        ty::Closure(..) |
-        ty::Generator(..) |
-        ty::Adt(..) |
-        // FIXME(eddyb) producing readable type names for trait objects can result
-        // in problematically distinct types due to HRTB and subtyping (see #47638).
+        // FIXME(eddyb) producing readable type names for trait objects can result in
+        // problematically distinct types due to HRTB and subtyping (see #47638).
         // ty::Dynamic(..) |
-        ty::Foreign(..) |
-        ty::Str => {
+        ty::Adt(..) | ty::Closure(..) | ty::Foreign(..) | ty::Generator(..) | ty::Str
+            if !cx.sess().fewer_names() =>
+        {
             let mut name = String::with_capacity(32);
             let printer = DefPathBasedNames::new(cx.tcx, true, true);
             printer.push_type_name(layout.ty, &mut name, false);
-            if let (&ty::Adt(def, _), &Variants::Single { index })
-                 = (&layout.ty.kind, &layout.variants)
+            if let (&ty::Adt(def, _), &Variants::Single { index }) =
+                (&layout.ty.kind, &layout.variants)
             {
                 if def.is_enum() && !def.variants.is_empty() {
                     write!(&mut name, "::{}", def.variants[index].ident).unwrap();
                 }
             }
-            if let (&ty::Generator(_, _, _), &Variants::Single { index })
-                 = (&layout.ty.kind, &layout.variants)
+            if let (&ty::Generator(_, _, _), &Variants::Single { index }) =
+                (&layout.ty.kind, &layout.variants)
             {
                 write!(&mut name, "::{}", ty::GeneratorSubsts::variant_name(index)).unwrap();
             }
             Some(name)
         }
-        _ => None
+        ty::Adt(..) => {
+            // If `Some` is returned then a named struct is created in LLVM. Name collisions are
+            // avoided by LLVM (with random renames). If rustc doesn't generate names then that
```

Comment:
s/random/[incrementing suffix](https://github.com/llvm/llvm-project/blob/6760f7ee6f00158c496b8e50d9ddac1f1c631ad7/llvm/lib/IR/Type.cpp#L435) (yes there's a "random rename" comment above but LLVM never actually injects randomness like that, it's more... colloquial use of "random" than RNG, let's say)

Reply:
Fixed!